### PR TITLE
Port gameplay changes from alumni night

### LINKED ISSUE #1614: Fix gameplay issues from testing with 2 AIs
### Description of the task

As titled, fix stuff we found while testing with 2 AIs

### Acceptance criteria

- [ ] as titled

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

This PR ports all gameplay fixes and improvements made while testing our AI against itself for alumni night.

### Testing Done

Updated unit tests, high-level gameplay validated at alumni night testing.

### Resolved Issues

resolves #1614 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/hl/stp/play/shoot_or_chip_play.cpp [hunk 1]
```diff
@@ -28,8 +28,11 @@ std::string ShootOrChipPlay::getName() const
 
 bool ShootOrChipPlay::isApplicable(const World &world) const
 {
-    return world.gameState().isPlaying() &&
-           teamHasPossession(world, world.friendlyTeam());
+    // TODO: figure out whether or not ShootOrChip and ShootOrPass should be separate
```
Comment (MathewMacDougall) [inline src/software/ai/hl/stp/play/shoot_or_chip_play.cpp]: Thoughts on this? I think they should be combined (perhaps in a future PR when we get to more strategy changes). In the meantime we can either leave them as-is, or temporarily delete ShootOrChip.

The main differences are that ShootOrChip will chip instead of pass when the robot with the ball is threatened by an enemy, and it uses findOpenAreas instead of cherry pickers to move robots into space.

File: src/software/simulated_tests/simulated_test_fixture.cpp [hunk 2]
```diff
@@ -27,7 +27,6 @@ void SimulatedTestFixture::SetUp()
     // guarantee the pointer will never be null / empty
     simulator = std::make_unique<Simulator>(Field::createSSLDivisionBField());
     ai = AI(DynamicParameters->getAIConfig(), DynamicParameters->getAIControlConfig());
-    sensor_fusion = SensorFusion(DynamicParameters->getSensorFusionConfig());
```
Comment (jonathanlew) [inline src/software/simulated_tests/simulated_test_fixture.cpp:30]: why was this deleted?

File: src/software/sensor_fusion/sensor_fusion.cpp [hunk 3]
```diff
@@ -156,15 +157,15 @@ void SensorFusion::updateWorld(const SSL_DetectionFrame &ssl_detection_frame)
     {
         for (auto &detection : ball_detections)
         {
-            invert(detection);
+            detection = invert(detection);
```
Comment (jonathanlew) [inline src/software/sensor_fusion/sensor_fusion.cpp:158]: assuming this fixes something, can we add a test for this?

File: src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp [hunk 4]
```diff
@@ -149,7 +149,7 @@ std::vector<ObstaclePtr> RobotNavigationObstacleFactory::createFromTeam(
 ObstaclePtr RobotNavigationObstacleFactory::createFromBallPosition(
     const Point &ball_position) const
 {
-    return createFromShape(Circle(ball_position, BALL_MAX_RADIUS_METERS));
+    return createFromShape(Circle(ball_position, BALL_MAX_RADIUS_METERS * 4));
```
Comment (jonathanlew) [inline src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp]: why was this changed?

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp [hunk 5]
```diff
@@ -35,146 +36,132 @@ void InterceptBallAction::accept(MutableActionVisitor& visitor)
...
-    // will move to intercept the ball at the point it would leave the field. This
-    // generally causes the robot to move far enough ahead of the ball that it can switch
-    // to intercepting at the closest point on the ray like above.
-    //
-    // Finally, if the ball is moving slowly the robot will go directly to the ball.
+        bool ball_very_close_to_robot_dribbler =
+            distance(ball.position(),
+                     robot->position() + Vector::createFromAngle(robot->orientation())
+                                             .normalize(DIST_TO_FRONT_OF_ROBOT_METERS)) <
+            0.2 + BALL_MAX_RADIUS_METERS;
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/action/intercept_ball_action.cpp]: are we making this a dynamic param?

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp [hunk 6]
```diff
@@ -35,146 +36,132 @@ void InterceptBallAction::accept(MutableActionVisitor& visitor)
...
+            Vector::createFromAngle(robot->orientation())
+                .perpendicular()
+                .normalize((DIST_TO_FRONT_OF_ROBOT_METERS + BALL_MAX_RADIUS_METERS) *
+                           robot->angularVelocity().toRadians());
+        Vector expected_ball_velocity = robot->velocity() + dribbler_tangential_velocity;
+        bool velocity_magnitudes_similar =
+            std::fabs(expected_ball_velocity.length() - ball.velocity().length()) < 0.05;
+        bool velocity_angles_similar =
+            expected_ball_velocity.orientation().minDiff(ball.velocity().orientation()) <
+            Angle::fromDegrees(10);
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/action/intercept_ball_action.cpp]: this should be a constant if not a dyanamic param

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp [hunk 7]
```diff
@@ -35,146 +36,132 @@ void InterceptBallAction::accept(MutableActionVisitor& visitor)
...
+                                             .normalize(DIST_TO_FRONT_OF_ROBOT_METERS)) <
+            0.2 + BALL_MAX_RADIUS_METERS;
+        Vector dribbler_tangential_velocity =
+            Vector::createFromAngle(robot->orientation())
+                .perpendicular()
+                .normalize((DIST_TO_FRONT_OF_ROBOT_METERS + BALL_MAX_RADIUS_METERS) *
+                           robot->angularVelocity().toRadians());
+        Vector expected_ball_velocity = robot->velocity() + dribbler_tangential_velocity;
+        bool velocity_magnitudes_similar =
+            std::fabs(expected_ball_velocity.length() - ball.velocity().length()) < 0.05;
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/action/intercept_ball_action.cpp]: 0.05 should also be a constant

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp [hunk 8]
```diff
@@ -35,146 +36,132 @@ void InterceptBallAction::accept(MutableActionVisitor& visitor)
...
-        }
-        else
-        {
-            ball_not_moving = true;
-        }
+        yield(std::make_unique<MoveIntent>(robot->id(), robot->position(),
+                                           face_ball_orientation, 0, 0,
+                                           DribblerEnable::ON, MoveType::NORMAL,
+                                           AutokickType::NONE, BallCollisionType::ALLOW));
+    } while (robot->velocity().length() > 0.03);
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/action/intercept_ball_action.cpp]: where does 0.03 come from?

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp [hunk 9]
```diff
@@ -35,146 +36,132 @@ void InterceptBallAction::accept(MutableActionVisitor& visitor)
...
 
-        std::optional<Point> intercept_pos = std::nullopt;
-        if ((point_in_front_of_ball &&
-             (ball_not_moving || ball_time_to_position > robot_time_to_pos)))
+        if (robot_time_to_pos < ball_time_to_position)
         {
-            intercept_pos = closest_point;
+            break;
         }
+        intercept_position += ball.velocity().normalize(0.1);
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/action/intercept_ball_action.cpp]: where does 0.1 come from?

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp [hunk 10]
```diff
@@ -35,146 +36,132 @@ void InterceptBallAction::accept(MutableActionVisitor& visitor)
...
 
+void InterceptBallAction::interceptFastBall(IntentCoroutine::push_type& yield)
+{
+    // Find the first point where the robot can get to before the ball
+    Point intercept_position = ball.position();
+    while (contains(field.fieldLines(), intercept_position))
+    {
+        Duration ball_time_to_position =
+            Duration::fromSeconds(distance(intercept_position, ball.position()) /
+                                  (ball.velocity().length() + 1e-6));
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/action/intercept_ball_action.cpp]: can we change 1e-6 to a constant or avoid it?

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp [hunk 11]
```diff
@@ -35,146 +36,132 @@ void InterceptBallAction::accept(MutableActionVisitor& visitor)
...
+    // Move to the intercept position and stop while avoiding the ball
+    bool robot_in_intercept_position = false;
+    do
+    {
+        yield(std::make_unique<MoveIntent>(robot->id(), intercept_position,
+                                           (-ball.velocity()).orientation(), 0, 0,
+                                           DribblerEnable::ON, MoveType::NORMAL,
+                                           AutokickType::NONE, BallCollisionType::AVOID));
+
+        if (ball.velocity().length() < BALL_MOVING_SLOW_SPEED_THRESHOLD / 2.0)
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/action/intercept_ball_action.cpp:126]: is BALL_MOVING_SLOW_SPEED_THRESHOLD 2 times larger than it should be?

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp [hunk 12]
```diff
@@ -35,146 +36,132 @@ void InterceptBallAction::accept(MutableActionVisitor& visitor)
...
+            restart();
         }
-        else
+
+        bool intercept_point_in_front_of_ball =
+            acuteAngle(ball.velocity(), intercept_position - ball.position()) <
+            Angle::quarter();
+        bool robot_at_intercept_point =
+            distance(robot->position(), intercept_position) < 0.015;
+        bool robot_stopped = robot->velocity().length() < 0.02;
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/action/intercept_ball_action.cpp]: where did we get these constants?

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp [hunk 13]
```diff
@@ -35,146 +36,132 @@ void InterceptBallAction::accept(MutableActionVisitor& visitor)
...
-                DribblerEnable::ON, MoveType::NORMAL, AutokickType::NONE,
-                BallCollisionType::ALLOW));
+            restart();
         }
-    } while (!robotHasPossession(ball.getPreviousStates(), robot->getPreviousStates()));
+        bool ball_very_close_to_robot_dribbler =
+            distance(ball.position(),
+                     robot->position() + Vector::createFromAngle(robot->orientation())
+                                             .normalize(DIST_TO_FRONT_OF_ROBOT_METERS)) <
+            0.2 + BALL_MAX_RADIUS_METERS;
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/action/intercept_ball_action.cpp]: yeah it should be dynamic param or constat. This is the second time I've seen it

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp [hunk 14]
```diff
@@ -35,146 +36,132 @@ void InterceptBallAction::accept(MutableActionVisitor& visitor)
...
+    bool intercept_done = false;
+    do
+    {
+        yield(std::make_unique<MoveIntent>(robot->id(), intercept_position,
+                                           (-ball.velocity()).orientation(), 0, 0,
+                                           DribblerEnable::ON, MoveType::NORMAL,
+                                           AutokickType::NONE, BallCollisionType::ALLOW));
+
+        if (!intercept_done &&
+            ball.velocity().length() < BALL_MOVING_SLOW_SPEED_THRESHOLD / 2.0)
```
Comment (jonathanlew) [inline src/software/ai/hl/stp/action/intercept_ball_action.cpp:151]: same as above for BALL_MOVING_SLOW_SPEED_THRESHOLD

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Lots of magic :mage: numbers :1234:...

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp [hunk 15]
```diff
@@ -35,146 +36,132 @@ void InterceptBallAction::accept(MutableActionVisitor& visitor)
...
     {
-        std::unordered_set<Point> intersections =
-            intersection(field.fieldLines(), ball_ray);
-        if (!intersections.empty())
+        auto face_ball_orientation = (ball.position() - robot->position()).orientation();
+        yield(std::make_unique<MoveIntent>(
+            robot->id(), ball.position(), face_ball_orientation, 0, 0, DribblerEnable::ON,
+            MoveType::NORMAL, AutokickType::NONE, BallCollisionType::ALLOW));
+
+        if (ball.velocity().length() > BALL_MOVING_SLOW_SPEED_THRESHOLD * 2)
```
Comment (garethellis0) [inline src/software/ai/hl/stp/action/intercept_ball_action.cpp:54]: Quick comment here on what this case represents physically would be helpful.

File: src/software/ai/hl/stp/play/shoot_or_chip_play.cpp (hunk 1 above)
Comment (garethellis0) [reply src/software/ai/hl/stp/play/shoot_or_chip_play.cpp]: We could just leave this as:
```
// NOTE: We do not currently use this play, as passing is generally superior. However we keep it around and
//              maintain it so as to have a backup if passing become unreliable for whatever reason
return true;
```
but I'm definitely on the fence about this one. If we keep it around we should make sure to have simulated tests for it to stop regressions.

**Review by garethellis0 — CHANGES_REQUESTED** [review]
Not much to add beyond Jon's comments.

**Review by akhilveeraghanta — COMMENTED** [review]
not much to add beyond existing comments either, really nice to see this come together

File: src/software/simulated_tests/simulated_test_fixture.cpp (hunk 2 above)
Comment (MathewMacDougall) [reply src/software/simulated_tests/simulated_test_fixture.cpp:30]: Due to the mutex added to SensorFusion, it no longer has a copy constructor so has to be created in the tests initialization list, and can't be copied in the setup here.

File: src/software/sensor_fusion/sensor_fusion.cpp (hunk 3 above)
Comment (MathewMacDougall) [reply src/software/sensor_fusion/sensor_fusion.cpp:158]: added test

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp (hunk 11 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/intercept_ball_action.cpp:126]: no the extra factor of 2 prevents the action from rapidly restarting if the ball speed is right at the threshold

File: src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp (hunk 4 above)
Comment (MathewMacDougall) [reply src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp]: We normally would perfectly brush up against any object we're avoiding, since we only account for the size of the object plus the robot radius. That value is then multiplied by a small buffer factor to get the total avoid distance / size of the obstacle. In the case of the ball, this meant the obstacle was much to small to actually avoid properly, so we were constantly running into the ball when we shouldn't have been.

My solution at the time was to increase the size of the ball obstacle, since we can't increase the "fudge factor" without affecting other obstacles.

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp (hunk 15 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/intercept_ball_action.cpp:54]: done

Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/intercept_ball_action.cpp:54]: done

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp (hunk 5 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/intercept_ball_action.cpp]: Started by making it a class constant since I'm not sure how much we'd want to tune it.

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp (hunk 6 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/intercept_ball_action.cpp]: made constant

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp (hunk 7 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/intercept_ball_action.cpp]: made constant

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp (hunk 8 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/intercept_ball_action.cpp]: somewhat arbitrary. made constant

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp (hunk 9 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/intercept_ball_action.cpp]: arbitrary. made constant

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp (hunk 10 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/intercept_ball_action.cpp]: avoided it since this case is only called when the ball is moving fast

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp (hunk 12 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/intercept_ball_action.cpp]: they're close to the MoveAction but made-up. Used proper constants.

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp (hunk 13 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/intercept_ball_action.cpp]: made constant

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp (hunk 14 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/intercept_ball_action.cpp:151]: see above

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp (hunk 11 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/intercept_ball_action.cpp:126]: I added a comment to both places this is done explaining.

**Comment by MathewMacDougall** [comment]
Resolved merge conflicts should be ready for re-review.

**Review by garethellis0 — DISMISSED** [review]
LGTM :+1: :tada:

File: src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp (hunk 4 above)
Comment (jonathanlew) [reply src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp]: Assuming you're talking about `RobotObstacleInflationFactor`, it multiplies ROBOT_MAX_RADIUS_METERS, not the radius of the obstacle, so if it's a problem for ball, it's probably a problem for other obstacles, too

File: src/software/sensor_fusion/sensor_fusion.h [hunk 16]
```diff
@@ -104,6 +104,7 @@ class SensorFusion
     std::optional<Ball> ball;
     Team friendly_team;
     Team enemy_team;
+    std::mutex game_state_mutex;
```
Comment (jonathanlew) [inline src/software/sensor_fusion/sensor_fusion.h]: shouldn't this be in ThreadedSensorFusion?

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp (hunk 11 above)
Comment (jonathanlew) [reply src/software/ai/hl/stp/action/intercept_ball_action.cpp:126]: don't like it much, but there's a comment explaining so I'll let it go

**Review by jonathanlew — COMMENTED** [review]
few more comments

File: src/software/sensor_fusion/sensor_fusion.h (hunk 16 above)
Comment (MathewMacDougall) [reply src/software/sensor_fusion/sensor_fusion.h]: Oh good point, I think this change may have initially been made before we split `SensorFusion` like that. I actually don't think we need the lock at all anymore. Now that SensorFusion just subscribes to a single buffer of `SensorMsg` we no longer have the case where different update functions can be called at the same time (due to SSL Vision and Refbox sending messages at the same time).

@garethellis0 sanity check me here?

File: src/software/ai/hl/stp/action/intercept_ball_action.cpp (hunk 11 above)
Comment (MathewMacDougall) [reply src/software/ai/hl/stp/action/intercept_ball_action.cpp:126]: I agree I'm not a huge fan, This will (99%) be replaced by the navigator and trajectory planner at some point :tm:

File: src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp (hunk 4 above)
Comment (MathewMacDougall) [reply src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp]: Do you have any ideas for a "more proper" fix here or is this fine for now?

Comment (jonathanlew) [reply src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp]: You can increase `RobotObstacleInflationFactor` to achieve the same result, i.e. solve for x: 0.203 = BALL_MAX_RADIUS_METERS + ROBOT_MAX_RADIUS_METERS * x. Then the new `RobotObstacleInflationFactor` is x

Comment (MathewMacDougall) [reply src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp]: New inflation factor would be 2.0167 by that calculation. This change causes a lot of other obstacle factory tests to fail (not unexpected). Although generally bigger obstacles seems desirable here based on the symptom and reason for the change, I'm not entirely sure that the inflation factor is the best way to address this problem.

It seems like fundamentally we make the "base" obstacle the exact shape of the physical object (eg. a circle of exactly the ball's radius), and then expand that based on the inflation factor. The scaling factor will inflate larger obstacles more than smaller obstacles, meaning if we make this change to make smaller obstacles "work" we may be over-expanding our larger obstacles. This is part of the reason I'm not a fan of using "scaling factors" here, but this is out of scope for this PR (happy to have offline discussion or open an issue where we can talk about this, just wanted to bring it up here).

Do you have a preference for changing the scaling factor vs the "base obstacle shape"? I'll do whichever is preferred for now since fundamentally changing too much else is out of scope here.

File: src/software/sensor_fusion/sensor_fusion.h (hunk 16 above)
Comment (MathewMacDougall) [reply src/software/sensor_fusion/sensor_fusion.h]: removed the mutex

File: src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp (hunk 4 above)
Comment (MathewMacDougall) [reply src/software/ai/navigator/obstacle/robot_navigation_obstacle_factory.cpp]: changed the scaling factor

File: src/software/sensor_fusion/sensor_fusion.h (hunk 16 above)
Comment (garethellis0) [reply src/software/sensor_fusion/sensor_fusion.h]: Seems sane to me.

**Review by jonathanlew — COMMENTED** [review]
LGTM pending CI

**Review by jonathanlew — DISMISSED** [review]

**Comment by MathewMacDougall** [comment]
Fixed tests please re-review / approve.

**Review by jonathanlew — DISMISSED** [review]

**Comment by MathewMacDougall** [comment]
Fixed merge conflicts again.

@LiCody @kelvinkoon @HaomiaoZ @EvanMorcom need review on this please.

**Comment by jonathanlew** [comment]
CI is failing because of nanopb converter bug. You can try restarting CI, wait for the fix to go in, or just pull these changes into this PR https://github.com/UBC-Thunderbots/Software/pull/1671

Edit: I meant #1677

**Comment by MathewMacDougall** [comment]
Ah. I'll just wait and restart CI in the meantime.

**Review by jonathanlew — APPROVED** [review]

**Review by kelvinkoon — APPROVED** [review]

**Comment by LiCody** [comment]
LGTM pending CI

**Review by LiCody — APPROVED** [review]
